DVCodingStyle.md

```diff
@@ -0,0 +1,864 @@
+3.  The `body()` method should only execute the raw functional behavior of
+    the sequence.
+    Any associated housekeeping code should be placed elsewhere, such as in the
+    `pre_start()` and `post_start()` methods.
+    *   `pre_start()` and `post_start()` must be used instead of `pre_body()`
+        and `post_body()`, as these will always be called, even for subsequences
+        called from a parent sequence.
+4.  When creating a sequence object, always randomize the sequence before starting it.
+5.  Virtual sequences must be used to coordinate the behavior of multiple agents and multiple sequences.
+6.  Virtual sequences must be started on the null sequencer or on a valid virtual sequencer.
+
+
+### Sequence Items
+
+It is recommended, but not necessary, to create small "unit tests" for
+transactions objects to ensure the randomization constraints are working as
+intended.
+
+
+### RAL Usage
+
+When using the UVM RAL model, care must be taken when using `uvm_reg::set()` and
+`uvm_reg::update()`, as these are both non-atomic operations that can easily
+cause race conditions if there are any parallel threads.
+
+
+### Objections and Coordinating End of Test
+
+1.  At a high level, a phase should not end when either there is more stimulus
+    to send during that phase or the DUT has yet to respond to some stimulus that
+    has been sent.
+    To prevent a phase from ending in the former case, raise and lower objections
+    inside tests.
+    To prevent a phase from ending in the latter case, raise objections in the
+    scoreboard whenever expect queues are not empty and lower them when they are empty.
+2.  It is recommended to avoid using objections in drivers or monitors
+    for performance reasons.
+3.  If a sequence needs to raise/drop objections, it should call
+    `raise_objection` in its `pre_start()` method and `drop_objection` in its
+    `post_start()` method. A good alternative to this is to call
+    `set_automatic_phase_objection(1)` in the sequence constructor.
+4.  Always perform the test `if (starting_phase != null)` before calling
+    `raise_objection` or `drop_objection` within a sequence. Prior to uvm-1.2,
+    `starting_phase` was a member of the class `uvm_sequence_base`.
+    From uvm-1.2 onward, the `starting_phase` variable is deprecated and instead
+    must be accessed using the `get_starting_phase` method.
+5.  When starting a sequence that can raise and drop objections, if you want
+    the sequence to raise and drop objections, set the `starting_phase` member of
+    the sequence object the `<sequence_handle>.set_starting_phase(<uvm_phase>)`
+    before starting the sequence.
+6.  To avoid reimplementing objection functionality in multiple sequence classes,
+    create a parent sequence that implements this functionality for its derivatives.
+7.  When calling `raise_objection` or `drop_objection`, you must pass a string
+    as a second argument to describe the objection to help with debugging.
```

Comment:
This is sensible, but can we give examples? In particular, it's a bit less clear what sensible strings for `drop_objection` should be.

Reply:
Sure, this has been added.